File: Problems/0538_Convert_BST_to_Greater_Tree/Project_CS/Operate_TreeNode.cs
using System;
using System.Collections.Generic;

public class Operate_TreeNode
{
    public TreeNode set_node(string[] flds, int depth, int pos)
    {
        if (flds.Length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.Pow(2, i);
        
        if (cur_pos + pos > flds.Length - 1)
            return null;
        
        if (flds[cur_pos + pos] == "null")
            return null;

        try
        {
            TreeNode node = new TreeNode(int.Parse(flds[cur_pos + pos]));
            node.left = set_node(flds, depth + 1, 2*pos);
            node.right = set_node(flds, depth + 1, 2*pos + 1);

            return node;
        }
        catch (Exception e)
        {
            Console.WriteLine("\n" +  e.Message + "\n" +
                              "set_node() Error ... flds[" + (cur_pos + pos).ToString() + "] = " + flds[cur_pos + pos] + "\n");
            Environment.Exit(-1);

            return null;
        }
    }

    List<string> resultStr;

    public string output(TreeNode node)
    {
        resultStr = new List<string>();

        string outStr = output_tree(node, 0);
        resultStr.Clear();

        return outStr;
    }

    private string output_tree(TreeNode node, int n)
    {
        if (node == null)
            return "";
        
        if (resultStr.Count <= n)
            resultStr.Add("(" + node.val.ToString() + ")");
        else
            resultStr[n] += ",(" + node.val.ToString() + ")";

        if (node.left != null)
            output_tree(node.left, n + 1);
        if (node.right != null)
            output_tree(node.right, n + 1);

        string outputStr = "";
        for (int i = 0; i < resultStr.Count; ++i)
        {
            outputStr += resultStr[i] + "\n";
        }

        return outputStr;
    }

    public string Tree2str(TreeNode t)
    {
        if (t == null)
            return "";

        string resultStr = t.val.ToString();

        if (t.left == null && t.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(t.left) + ")";
        if (t.right != null)
            resultStr += "(" + Tree2str(t.right) + ")";

        return resultStr;
    }
}

File: Problems/0100_0199/0100_Same_Tree/Project_Elixir/lib/OperateTreeNode.ex
defmodule OperateTreeNode do
  @spec createTreeNode(nums :: [String.t]) :: TreeNode.t | nil
  def createTreeNode(nums) do
    createTreeNode(nums, 0, 0)
  end

# @spec createTreeNode(nums :: [String.t], depth :: Integer, pos :: Integer) :: TreeNode.t | nil
  def createTreeNode(nums, depth, pos) do
    cond do
      pos >= Enum.count(nums) ->
        nil
      Enum.count(nums) <= 0 ->
        nil
      true ->
        cur_pos = get_cur_pos(depth, 0)
        if (cur_pos + pos) > (Enum.count(nums) - 1) do
          nil
        else
          if Enum.at(nums, cur_pos + pos) == "null" do
            nil
          else
            %TreeNode{val: String.to_integer(Enum.at(nums, cur_pos + pos)), left: createTreeNode(nums, depth + 1, 2*pos), right: createTreeNode(nums, depth + 1, 2*pos + 1)}
          end
        end
    end
  end

  @spec get_cur_pos(depth :: Integer, i :: Integer) :: Integer
  def get_cur_pos(depth, i) do
    if i < depth do
      (:math.pow(2, i) |> round) + get_cur_pos(depth, i + 1)
    else
      0
    end
  end

  @spec treenode_to_string(node :: TreeNode.t | nil) :: String.t
  def treenode_to_string(node) do
    cond do
      node == nil ->
        ""
      node.left != nil and node.right != nil ->
        Integer.to_string(node.val) <> "(" <> treenode_to_string(node.left) <> ")(" <> treenode_to_string(node.right) <> ")"
      node.left != nil ->
        Integer.to_string(node.val) <> "(" <> treenode_to_string(node.left) <> ")()"
      node.right != nil ->
        Integer.to_string(node.val) <> "()(" <> treenode_to_string(node.right) <> ")"
      true ->
        Integer.to_string(node.val)
    end
  end

  @spec treenode_to_string2(node :: TreeNode.t | nil) :: String.t
  def treenode_to_string2(node) do
    if node == nil do
      ""
    else
      if node.left != nil and node.right != nil do
        Integer.to_string(node.val) <> "(" <> treenode_to_string(node.left) <> ")(" <> treenode_to_string(node.right) <> ")"
      else
        if node.left != nil do
          Integer.to_string(node.val) <> "(" <> treenode_to_string(node.left) <> ")()"
        else
          if node.right != nil do
            Integer.to_string(node.val) <> "()(" <> treenode_to_string(node.right) <> ")"
          else
            Integer.to_string(node.val)
          end
        end
      end
    end
  end
end
